---
---

<center><h1> Assignment </h1></center>

---
---

#### `Instructions`

* ***You are free to use internet to refer the documentation of any library.***
* ***If required, feel free to import other libraries to do the task required.***
* ***Before making a submission, make sure that your code is commented.***

---


#### `Importing the Required Libraries`

---

In [2]:
import pandas as pd
import numpy  as np
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

### Reading the Data

In [3]:
data = pd.read_csv('telecom_churn.csv')

### Q 1: Have a first look at the data by observing the followings :-
* Head of the data
* Shape of the data
* Missing Values in the data 
* Datatype of all features in the data
* Identify continuous features in the data
* Identify categorical features in the data
* Analyse descriptive statistics in the data

In [4]:
### Head of the data
data.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

In [5]:
### Shape of the data
data.shape

(3333, 20)

In [6]:
### Missing Values in the data
data.isnull().sum().sum()

0

In [7]:
### Datatype of all features in the data
data.dtypes

State                      object
Account length              int64
Area code                   int64
International plan         object
Voice mail plan            object
Number vmail messages       int64
Total day minutes         float64
Total day calls             int64
Total day charge          float64
Total eve minutes         float64
Total eve calls             int64
Total eve charge          float64
Total night minutes       float64
Total night calls           int64
Total night charge        float64
Total intl minutes        float64
Total intl calls            int64
Total intl charge         float64
Customer service calls      int64
Churn                        bool
dtype: object

In [8]:
# Identify continuous features and categorical features in the data
def identify_continuous_features(data):
     df=pd.DataFrame(data.nunique(),columns=['n_unique_values'])
     df['dtypes']=data.dtypes
     print(df)
     discrete_cols=data.select_dtypes(include=['int64']).columns
     continuous_cols=data.select_dtypes(include=['float64']).columns
     categorical_cols=data.select_dtypes(include=['object','bool']).columns
     print('discrete_cols:\n {} \ncontinuous_cols: \n{} \ncategorical_cols:\n {}'.format(discrete_cols,continuous_cols,categorical_cols))
     return discrete_cols,continuous_cols,categorical_cols
discrete_cols,continuous_cols,categorical_cols=identify_continuous_features(data)

                        n_unique_values   dtypes
State                                51   object
Account length                      212    int64
Area code                             3    int64
International plan                    2   object
Voice mail plan                       2   object
Number vmail messages                46    int64
Total day minutes                  1667  float64
Total day calls                     119    int64
Total day charge                   1667  float64
Total eve minutes                  1611  float64
Total eve calls                     123    int64
Total eve charge                   1440  float64
Total night minutes                1591  float64
Total night calls                   120    int64
Total night charge                  933  float64
Total intl minutes                  162  float64
Total intl calls                     21    int64
Total intl charge                   162  float64
Customer service calls               10    int64
Churn               

1.All int and float data type features are apparently numeric except 'Area code'

(We have few  unique values in columns:'Account length' and
  'Customer service calls',but 'Account length' can be understood as the number of days that
  particular account has been active and 'Customer service calls' the number of
  calls placed to Customer Service)
  
2.All object and bool data types features are considered as categorical.
'Area code' which is int data type is also categorical feature

In [9]:
discrete_columns=list(discrete_cols)
discrete_columns.remove('Area code')
print("Discrete features are \n {}\nTotal number of Discrete features:{}".format(discrete_columns,len(discrete_columns)))

Discrete features are 
 ['Account length', 'Number vmail messages', 'Total day calls', 'Total eve calls', 'Total night calls', 'Total intl calls', 'Customer service calls']
Total number of Discrete features:7


In [10]:
# Identify categorical features in the data
categorical_columns=list(categorical_cols)
categorical_columns.append('Area code')
print("categorical features are \n {}\nTotal number of categorical features:{}".format(categorical_columns,len(categorical_columns)))

categorical features are 
 ['State', 'International plan', 'Voice mail plan', 'Churn', 'Area code']
Total number of categorical features:5


In [11]:
#Identify continuous features in the data
print("Continuous features are \n {}\nTotal number of Continuous features:{}".format(continuous_cols,len(continuous_cols)))

Continuous features are 
 Index(['Total day minutes', 'Total day charge', 'Total eve minutes',
       'Total eve charge', 'Total night minutes', 'Total night charge',
       'Total intl minutes', 'Total intl charge'],
      dtype='object')
Total number of Continuous features:8


In [12]:
### Analyse descriptive statistics in the data

#limit  decimal values 
pd.set_option('float_format', '{:.1f}'.format)

# Find data statistics using describe function
describe_numeric=data.describe().T

# remove 'count' column in describe data frame as there are no null values   
describe_numeric.drop('count',axis=1,inplace=True)

In [13]:
describe_numeric

mean  std   min   25%   50%   75%   max
Account length         101.1 39.8   1.0  74.0 101.0 127.0 243.0
Area code              437.2 42.4 408.0 408.0 415.0 510.0 510.0
Number vmail messages    8.1 13.7   0.0   0.0   0.0  20.0  51.0
Total day minutes      179.8 54.5   0.0 143.7 179.4 216.4 350.8
Total day calls        100.4 20.1   0.0  87.0 101.0 114.0 165.0
Total day charge        30.6  9.3   0.0  24.4  30.5  36.8  59.6
Total eve minutes      201.0 50.7   0.0 166.6 201.4 235.3 363.7
Total eve calls        100.1 19.9   0.0  87.0 100.0 114.0 170.0
Total eve charge        17.1  4.3   0.0  14.2  17.1  20.0  30.9
Total night minutes    200.9 50.6  23.2 167.0 201.2 235.3 395.0
Total night calls      100.1 19.6  33.0  87.0 100.0 113.0 175.0
Total night charge       9.0  2.3   1.0   7.5   9.1  10.6  17.8
Total intl minutes      10.2  2.8   0.0   8.5  10.3  12.1  20.0
Total intl calls         4.5  2.5   0.0   3.0   4.0   6.0  20.0
Total intl charge        2.8  0.8   0.0   2.3   2.8   3.3   5.4
Customer service calls   1.6  1.3   0.0   1.0   1.0   2.0   9.0

1.On an average , the accounts of customers are active for 101.1 days 
  and no customer's account is active more than 243 days

2.75 percent of customers send messages through vmail below 20 times.
   and only 25 percent of customers send messages through 
   vmail for 20 to 51 times

3. Charges are apparently low at night than that of charges during the day
 (Even though the total number of day calls & 
night calls , day minutes & night minutes are almost similar in average & 
quantile distributions)

4.Average number of calls is ~100 whether its day/evening/night

5.No customer with zero total calls at night

6.Averages of  total day minutes and total day charges are 179.8 , 30.6 respectively

7.Only 25% of people make more than 114 total day calls(upto 165).

8.Averages of  total evening minutes and total evenings charges are 201.0 , 17.1 respectively

9.Only 25% of people make more than 114 total evening calls(upto 170).

10.Averages of  total night minutes and total night charges are 200.9 , 9 respectively

11.Only 25% of people make more than 113 total night calls(upto 175).

12.75% of customers make less than 6 'Total intl calls', maximum 
international calls made is 20.

13.Highest Total internation charge: 5.4
    
14.Only 25% of customers make more than 2 'Customer service calls'(upto 9)

In [14]:
#convert 'Area_code' feature into category data type 
#convert'object','bool' data type features into category
for i in categorical_columns:
     data[i]=data[i].astype("category")
        
        
describe_cats=data.describe(include=['category']).T
# remove 'count' column in describe_cats data frame as there are no null values   
describe_cats.drop('count',axis=1,inplace=True)
describe_cats

unique    top  freq
State                  51     WV   106
Area code               3    415  1655
International plan      2     No  3010
Voice mail plan         2     No  2411
Churn                   2  False  2850

In [15]:
pd.set_option('float_format', '{:.3f}'.format)
for i in data.select_dtypes(include='category'):
      print(data[i].value_counts(normalize='True'))
      print('\n')
        

WV   0.032
MN   0.025
NY   0.025
AL   0.024
OR   0.023
WI   0.023
OH   0.023
VA   0.023
WY   0.023
CT   0.022
VT   0.022
ID   0.022
MI   0.022
UT   0.022
TX   0.022
IN   0.021
KS   0.021
MD   0.021
NJ   0.020
NC   0.020
MT   0.020
CO   0.020
NV   0.020
WA   0.020
RI   0.020
MS   0.020
MA   0.020
AZ   0.019
MO   0.019
FL   0.019
ND   0.019
ME   0.019
NM   0.019
OK   0.018
NE   0.018
DE   0.018
SC   0.018
SD   0.018
KY   0.018
IL   0.017
NH   0.017
AR   0.017
DC   0.016
GA   0.016
HI   0.016
TN   0.016
AK   0.016
LA   0.015
PA   0.014
IA   0.013
CA   0.010
Name: State, dtype: float64


415   0.497
510   0.252
408   0.251
Name: Area code, dtype: float64


No    0.903
Yes   0.097
Name: International plan, dtype: float64


No    0.723
Yes   0.277
Name: Voice mail plan, dtype: float64


False   0.855
True    0.145
Name: Churn, dtype: float64




1.Company has customers from 51 unique states, 
'WV' state has high percentage of customers comparatively i.e 3%

2.Almost 50% of the customers are from 'Area Code' 415

3.90% of customers are not opted for international plan

4.~30% of customers subscribed for voice mail plan

5.~14% of customers are churning.

In [16]:
data['International plan'].value_counts(normalize='True')

No    0.903
Yes   0.097
Name: International plan, dtype: float64

In [17]:
data['Area code'].value_counts(normalize='True')

415   0.497
510   0.252
408   0.251
Name: Area code, dtype: float64

In [18]:
data['Churn'].value_counts(normalize='True')

False   0.855
True    0.145
Name: Churn, dtype: float64

### Q 2 : Convert the datatype of Churn variable into integer? What is the proportion of churning customer ?

In [19]:
data['Churn']=data['Churn'].astype('int')

In [20]:
m=data['Churn'].value_counts(normalize=True)
print('Proportion of churning customers: {:.2f}%'.format(m[0]*100))

Proportion of churning customers: 85.51%


In [21]:
k=data['Churn'][data['Churn']==0].count()/data['Churn'][data['Churn']==1].count()
print("proportion ratio \nchurn_positive:churn_negative= {:.2f}".format(k))

proportion ratio 
churn_positive:churn_negative= 5.90


### Q 3 : Select the customers coming from state 'TN', identify the churning rate of such customers ?

In [22]:
group_by_state=data.groupby('State')

In [23]:
m=group_by_state.get_group("TN")['Churn'].value_counts(normalize=True)
m

0   0.906
1   0.094
Name: Churn, dtype: float64

In [24]:
print("Churing rate from state'TN':{:.2f}%".format(m[1]*100))

Churing rate from state'TN':9.43%


### Q 4:  Select the data where the customers' account length  is multiple of 12? 

In [25]:
data[data['Account length']%12==0]

State  Account length Area code International plan Voice mail plan  \
3       OH              84       408                Yes              No   
12      IA             168       408                 No              No   
24      VA             132       510                 No              No   
33      AZ              12       408                 No              No   
35      GA              72       415                 No             Yes   
...    ...             ...       ...                ...             ...   
3284    AZ              48       415                 No             Yes   
3301    CA              84       415                 No              No   
3321    VT              60       415                 No              No   
3326    OH              96       415                 No              No   
3328    AZ             192       415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
3                         0            299.400               71   
12                        0            128.800               96   
24                        0             81.100               86   
33                        0            249.600              118   
35                       37            220.000               80   
...                     ...                ...              ...   
3284                     27            141.100              109   
3301                      0            280.000              113   
3321                      0            193.900              118   
3326                      0            106.600              128   
3328                     36            156.200               77   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3               50.900             61.900               88             5.260   
12              21.900            104.900               71             8.920   
24              13.790            245.200               72            20.840   
33              42.430            252.400              119            21.450   
35              37.400            217.300              102            18.470   
...                ...                ...              ...               ...   
3284            23.990            224.700               94            19.100   
3301            47.600            202.200               90            17.190   
3321            32.960             85.000              110             7.230   
3326            18.120            284.800               87            24.210   
3328            26.550            215.500              126            18.320   

      Total night minutes  Total night calls  Total night charge  \
3                 196.900                 89               8.860   
12                141.100                128               6.350   
24                237.000                115              10.670   
33                280.200                 90              12.610   
35                152.800                 71               6.880   
...                   ...                ...                 ...   
3284              174.300                122               7.840   
3301              156.800                103               7.060   
3321              210.100                134               9.450   
3326              178.900                 92               8.050   
3328              279.100                 83              12.560   

      Total intl minutes  Total intl calls  Total intl charge  \
3                  6.600                 7              1.780   
12                11.200                 2              3.020   
24                10.300                 2              2.780   
33                11.800                 3              3.190   
35                14.700                 6              3.970   
...                  ...               ...                ...   
3284              13.200                 2              3.560   
3

### Q 5: Select the data where State's name beings with the letter W?

In [26]:
data[data['State'].str.startswith('W')]

State  Account length Area code International plan Voice mail plan  \
9       WV             141       415                Yes             Yes   
26      WY              57       408                 No             Yes   
44      WI              64       510                 No              No   
49      WY              97       415                 No             Yes   
54      WY              87       415                 No              No   
...    ...             ...       ...                ...             ...   
3278    WI              87       415                 No              No   
3303    WI             114       415                 No             Yes   
3319    WY              89       415                 No              No   
3324    WV             159       415                 No              No   
3329    WV              68       415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
9                        37            258.600               84   
26                       39            213.000              115   
44                        0            154.000               67   
49                       24            133.200              135   
54                        0            151.000               83   
...                     ...                ...              ...   
3278                      0            238.000               97   
3303                     26            137.100               88   
3319                      0            115.400               99   
3324                      0            169.800              114   
3329                      0            231.100               57   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
9               43.960            222.000              111            18.870   
26              36.210            191.100              112            16.240   
44              26.180            225.800              118            19.190   
49              22.640            217.200               58            18.460   
54              25.670            219.700              116            18.670   
...                ...                ...              ...               ...   
3278            40.460            164.500               97            13.980   
3303            23.310            155.700              125            13.230   
3319            19.620            209.900              115            17.840   
3324            28.870            197.700              105            16.800   
3329            39.290            153.400               55            13.040   

      Total night minutes  Total night calls  Total night charge  \
9                 326.400                 97              14.690   
26                182.700                115               8.220   
44                265.300                 86              11.940   
49                 70.600                 79               3.180   
54                203.900                127               9.180   
...                   ...                ...                 ...   
3278              282.500                132              12.710   
3303              247.600                 94              11.140   
3319              280.900                112              12.640   
3324              193.700                 82               8.720   
3329              191.300                123               8.610   

      Total intl minutes  Total intl calls  Total intl charge  \
9                 11.200                 5              3.020   
26                 9.500                 3              2.570   
44                 3.500                 3              0.950   
49                11.000                 3              2.970   
54                 9.700                 3              2.620   
...                  ...               ...                ...   
3278              10.600                 6              2.860   
3

### Q 6: Convert the datatype International plan variable into boolean?

In [27]:
data['International plan'].astype('bool')

0       True
1       True
2       True
3       True
4       True
        ... 
3328    True
3329    True
3330    True
3331    True
3332    True
Name: International plan, Length: 3333, dtype: bool

### Q 7: Generate additional feature named F_X such that :-
* International plan is True then the value of F_X is 1
* If International plan is False and Account Length is more than the average Account length then F_X is 1
* If Number vmail messages is greater than 75th percentile and Customer Service calls is greater than 1 then F_X is 1
* In all other cases F_X is 0

In [28]:
data.columns

Index(['State', 'Account length', 'Area code', 'International plan',
       'Voice mail plan', 'Number vmail messages', 'Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge', 'Customer service calls',
       'Churn'],
      dtype='object')

In [29]:
data['F_X']=0
for i,j in enumerate(data['F_X']):
    v=data['Number vmail messages'][i]>data['Number vmail messages'].quantile(0.75)
    w=data['Customer service calls'][i]>1
    if data['International plan'][i]==True:
        data['F_X'][i]=1
    elif data['Account length'][i]>data['Account length'].mean():
        data['F_X'][i]=1
    elif v==True and w==True:
        data['F_X'][i]=1  

In [30]:
data['F_X'].value_counts()

1    1808
0    1525
Name: F_X, dtype: int64

### Q 8 : Research about 10 intersting pandas/numpy function and show their implementation in jupyter notebook. 

### 1 pd.cut ()

In [31]:
data['Customer service calls'].value_counts()

1    1181
2     759
0     697
3     429
4     166
5      66
6      22
7       9
8       2
9       2
Name: Customer service calls, dtype: int64

In [32]:
data['Customer_service_calls_bins'] = pd.cut(data['Customer service calls'], 
                                         bins = [-np.inf, 0, 2, np.inf], labels = ['0', 'less_than_2', 'greater_than_2'])

In [33]:
data['Customer_service_calls_bins'].value_counts()

less_than_2       1940
0                  697
greater_than_2     696
Name: Customer_service_calls_bins, dtype: int64

### 2. memory_usage()

memory_usage() returns a Pandas Series having the memory usage of each column
(in bytes) in a Pandas DataFrame. 

In [34]:
data.memory_usage(deep=True)

Index                            128
State                           8446
Account length                 26664
Area code                       3465
International plan              3560
Voice mail plan                 3560
Number vmail messages          26664
Total day minutes              26664
Total day calls                26664
Total day charge               26664
Total eve minutes              26664
Total eve calls                26664
Total eve charge               26664
Total night minutes            26664
Total night calls              26664
Total night charge             26664
Total intl minutes             26664
Total intl calls               26664
Total intl charge              26664
Customer service calls         26664
Churn                          13332
F_X                            26664
Customer_service_calls_bins     3638
dtype: int64

### 3.sort_values()

In [35]:
data.sort_values(by='Account length')

State  Account length Area code International plan Voice mail plan  \
923     MS               1       415                 No              No   
1838    IA               1       408                 No             Yes   
1490    TN               1       415                 No              No   
2960    TX               1       415                 No              No   
2423    SC               1       408                 No              No   
...    ...             ...       ...                ...             ...   
1886    ID             224       510                 No              No   
1751    WY             225       415                 No              No   
1551    MI             225       415                 No              No   
1408    NM             232       408                 No              No   
817     UT             243       510                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
923                       0            144.800              107   
1838                     26            208.000              115   
1490                      0            196.100              107   
2960                      0            182.100              106   
2423                      0            123.800              113   
...                     ...                ...              ...   
1886                      0            111.400              133   
1751                      0            182.700              142   
1551                      0            165.400              106   
1408                      0            165.600              104   
817                       0             95.500               92   

      Total day charge  Total eve minutes  ...  Total night minutes  \
923             24.620            112.500  ...              218.700   
1838            35.360            185.000  ...              177.700   
1490            33.340            296.500  ...              211.500   
2960            30.960            134.900  ...              152.300   
2423            21.050            236.200  ...               73.200   
...                ...                ...  ...                  ...   
1886            18.940            175.000  ...              217.200   
1751            31.060            246.500  ...              218.000   
1551            28.120            273.700  ...              210.000   
1408            28.150            195.900  ...              118.300   
817             16.240            163.700  ...              264.200   

      Total night calls  Total night charge  Total intl minutes  \
923                  79               9.840              13.800   
1838                144               8.000               8.100   
1490                 91               9.520               7.000   
2960                 75               6.850              10.000   
2423                 81               3.290               3.700   
...                 ...                 ...                 ...   
1886                106               9.770               5.500   
1751                103               9.810               8.800   
1551                 93               9.450               8.700   
1408                 77               5.320              11.800   
817                 118              11.890               6.600   

      Total intl calls  Total intl charge  Customer service calls  Churn  F_X  \
923                  3              3.730                       1      0    0   
1838                 9              2.190                       1      0    0   
1490                 2              1.890                       1      0    0   
2960                 3              2.700                       5      1    0   
2423                 2              1.000                       0      0    0   
...                ...                ...                     ...    ...  ...   
1886                 6              1.490                       3      0  

### 4.pipe()

pipe function organizes multiple preprocessing operations into a single one

In [36]:
def fill_missing_values(df):
   for col in df.select_dtypes(include= ["int","float"]).columns:
      val = df[col].mean()
      df[col].fillna(val, inplace=True)
   return df
def remove_outliers(df, column_list):
   for col in column_list:
      avg = df[col].mean()
      std = df[col].std()
      low = avg - 2 * std
      high = avg + 2 * std
      df = df[df[col].between(low, high, inclusive=True)]
   return df


In [37]:
data1 = (data.pipe(fill_missing_values).
                      pipe(remove_outliers, ['Total day minutes', 'Total day charge', 'Total eve minutes',
       'Total eve charge', 'Total night minutes', 'Total night charge',
       'Total intl minutes', 'Total intl charge']))

In [38]:
data1['Total night minutes'].nunique()

1250

In [39]:
data['Total night minutes'].nunique()

1591

### 5.explode()

In [40]:
df = pd.Series([1, 2, 3, [4, 5, 6], 7, [8, 9, 10]]).to_frame("s")
df

s
0           1
1           2
2           3
3   [4, 5, 6]
4           7
5  [8, 9, 10]

In [41]:
df.explode("s", ignore_index=True)

s
0   1
1   2
2   3
3   4
4   5
5   6
6   7
7   8
8   9
9  10

### 6.style()

In [42]:
data.describe().T.drop("count", axis=1)\
                 .style.highlight_max(color="darkred")

In [43]:
data.describe().T.drop("count", axis=1).style.background_gradient(
    subset=["mean", "50%"], cmap="Reds"
)

### 7.nlargest()  &  nsmallest()

In [44]:
data.nlargest(5, "Account length")

State  Account length Area code International plan Voice mail plan  \
817     UT             243       510                 No              No   
1408    NM             232       408                 No              No   
1551    MI             225       415                 No              No   
1751    WY             225       415                 No              No   
416     DE             224       510                Yes              No   

      Number vmail messages  Total day minutes  Total day calls  \
817                       0             95.500               92   
1408                      0            165.600              104   
1551                      0            165.400              106   
1751                      0            182.700              142   
416                       0            171.500               99   

      Total day charge  Total eve minutes  ...  Total night minutes  \
817             16.240            163.700  ...              264.200   
1408            28.150            195.900  ...              118.300   
1551            28.120            273.700  ...              210.000   
1751            31.060            246.500  ...              218.000   
416             29.160            160.000  ...              212.400   

      Total night calls  Total night charge  Total intl minutes  \
817                 118              11.890               6.600   
1408                 77               5.320              11.800   
1551                 93               9.450               8.700   
1751                103               9.810               8.800   
416                 102               9.560               5.000   

      Total intl calls  Total intl charge  Customer service calls  Churn  F_X  \
817                  6              1.780                       2      0    1   
1408                 3              3.190                       1      0    1   
1551                 3              2.350                       0      1    1   
1751                 2              2.380                       1      0    1   
416                  2              1.350                       1      1    1   

      Customer_service_calls_bins  
817                   less_than_2  
1408                  less_than_2  
1551                            0  
1751                  less_than_2  
416                   less_than_2  

[5 rows x 22 columns]

In [45]:
data.nsmallest(5, "Account length")

State  Account length Area code International plan Voice mail plan  \
923     MS               1       415                 No              No   
1409    NJ               1       415                 No             Yes   
1490    TN               1       415                 No              No   
1838    IA               1       408                 No             Yes   
2423    SC               1       408                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
923                       0            144.800              107   
1409                     30            183.100               95   
1490                      0            196.100              107   
1838                     26            208.000              115   
2423                      0            123.800              113   

      Total day charge  Total eve minutes  ...  Total night minutes  \
923             24.620            112.500  ...              218.700   
1409            31.130            232.600  ...              248.300   
1490            33.340            296.500  ...              211.500   
1838            35.360            185.000  ...              177.700   
2423            21.050            236.200  ...               73.200   

      Total night calls  Total night charge  Total intl minutes  \
923                  79               9.840              13.800   
1409                110              11.170               8.400   
1490                 91               9.520               7.000   
1838                144               8.000               8.100   
2423                 81               3.290               3.700   

      Total intl calls  Total intl charge  Customer service calls  Churn  F_X  \
923                  3              3.730                       1      0    0   
1409                 2              2.270                       0      0    0   
1490                 2              1.890                       1      0    0   
1838                 9              2.190                       1      0    0   
2423                 2              1.000                       0      0    0   

      Customer_service_calls_bins  
923                   less_than_2  
1409                            0  
1490                  less_than_2  
1838                  less_than_2  
2423                            0  

[5 rows x 22 columns]

### 8.sample()

In [46]:
data.sample(n=5)

State  Account length Area code International plan Voice mail plan  \
2847    ID             119       415                Yes             Yes   
1122    NE             159       415                 No              No   
1666    KY              88       415                 No              No   
2810    NH              90       415                 No             Yes   
1656    ND             101       415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
2847                     32            173.000              101   
1122                      0            189.100              105   
1666                      0            148.200               82   
2810                     42            193.300               66   
1656                     28            105.900              132   

      Total day charge  Total eve minutes  ...  Total night minutes  \
2847            29.410            209.400  ...              231.100   
1122            32.150            246.100  ...              242.000   
1666            25.190            308.700  ...              235.400   
2810            32.860            263.300  ...              214.400   
1656            18.000            231.700  ...              281.300   

      Total night calls  Total night charge  Total intl minutes  \
2847                 91              10.400              12.200   
1122                106              10.890              10.400   
1666                 79              10.590               6.400   
2810                 97               9.650              11.100   
1656                120              12.660              10.700   

      Total intl calls  Total intl charge  Customer service calls  Churn  F_X  \
2847                 4              3.290                       0      0    1   
1122                 5              2.810                       1      1    1   
1666                 4              1.730                       2      0    0   
2810                 4              3.000                       0      0    0   
1656                 5              2.890                       1      0    0   

      Customer_service_calls_bins  
2847                            0  
1122                  less_than_2  
1666                  less_than_2  
2810                            0  
1656                  less_than_2  

[5 rows x 22 columns]

### 9 Query()

In [47]:
df = pd.DataFrame({'A': range(1, 6),
                   'B': range(10, 0, -2),
                   'C C': range(10, 5, -1)})

In [48]:
df.query('A > B')

A  B  C C
4  5  2    6

### 10.argpartition()

argpartition() is use to find N largest values index.
The output will be the N largest values index, 
and then we can sort the values if needed.

In [49]:
x = np.array([12, 10, 12, 0, 6, 8, 9, 1, 16, 4, 6, 0])
index_val = np.argpartition(x, 5)[5:]
index_val


array([10,  5,  6,  8,  1,  2,  0], dtype=int64)

In [50]:
np.sort(x[index_val])

array([ 6,  8,  9, 10, 12, 12, 16])